logverification/verifyconsistency_test.go

```diff
@@ -0,0 +1,130 @@
+// A user wanting to convince themselves that DataTrails faithfully maintains the long-term
+// integrity of their data might do something like this:
+//  0. Keep backups of the merkle log data for their tenancy, to check that previous entries
+//     are still in the latest version of the log.
+//  1. Fetch the current signed state of the log which includes entries appended since (0)
+//  2. Checking the *consistency* of the new signed log state against the state you're confident in
+//     means that every node (leaves and intermediate hash nodes) are included in the new state,
+//     and at the same positions, and you can trust this as the new head of the log.
+//  3. Get the hashes of peaks in your trusted log state and verify using a consistency proof based
+//     on a published signed state.
+func TestVerifyConsistency(t *testing.T) {
+	testContext, testGenerator, _ := integrationsupport.NewAzuriteTestContext(t, "TestVerifyConsistency")
+
+	// - Public key to verify the signature of the signed state (ephemeral keys are used for this demo)
+	// - Massif context that represents the backed-up merkle log data for a point in the past (containing 7 events)
+	// - Massif context that represents the head of the log (after 4 additional events appended)
+	//
+	// (!) A MassifContext provides access to the log entries in a single massif. A massif is a
+	// logical chunk of the merkle log, stored in an Azure blob.
+	publicVerificationKey, backupMassifContext, newMassifContext := integrationsupport.SetupTest(t, testContext, testGenerator)
+
+	hasher := sha256.New()
+	codec, err := massifs.NewRootSignerCodec()
+	require.NoError(t, err)
+
+	// First we verify the signature. In order to trust the data, we want to see that DataTrails has
+	// committed to the state of the merkle log by signing it.
+	logStateNow, signatureVerificationErr := verifySignature(
+		context.Background(), publicVerificationKey, hasher, codec, testContext.Storer,
+		&newMassifContext,
+	)
+	require.NoError(t, signatureVerificationErr)
+
+	// Then we verify the consistency of the data between our previously trusted log state
+	// (accessible through backupMassifContext) and the current state (accessible through
+	// newMassifContext.)
+
+	// backupMassifContext points to the following merkle log with 7 leaves.
+	//
+	//	    6
+	//	  /   \
+	//	 2     5     9
+	//	/ \   / \   / \
+	// 0   1 3   4 7   8 10   <- Leaf Nodes
+
+	// newMassifContext points to the following merkle log with 11 leaves. You can see the exact
+	// structure of backupMassifContext within it (ending at leaf mmr index 10.) This illustrates
+	// what it means for the merkle trees to be consistent.
+	//
+	//	         14
+	//	        /  \
+	//	       /    \
+	//	      /      \
+	//	     /        \
+	//	    6          13
+	//	  /   \       /   \
+	//	 2     5     9     12     17
+	//	/ \   / \   / \   /  \   /  \
+	// 0   1 3   4 7   8 10  11 15  16  18 <- Leaf Nodes
+	verified, err := VerifyConsistencyFromMassifs(
```

Comment:
If its possible we should move away from `VerifyConsistencyFromMassifs`? i think the demos use `VerifyConsistency` which takes 2 log states instead of 2 massifs

Reply:
Quite possibly yes - This was just the least work to migrate the tests over. I've still got to add depth here, but we're working from generated data so I'd have to reconstruct the full state anyway.

Reply:
yeh thats a good first step :) happy to resolve

Reply:
Fixforward